xc/device.ml

```diff
@@ -1777,8 +1781,33 @@ let __start (task: Xenops_task.task_handle) ~xs ~dmpath ?(timeout = !Xenopsd.qem
 	let cancel = Cancel_utils.Qemu (qemu_domid, domid) in
 	let qemu_pid = init_daemon ~task ~path:dmpath ~args ~name:"qemu-dm" ~domid
 		~xs ~ready_path ~ready_val:"running" ~timeout ~cancel () in
+	match !Xenopsd.action_after_qemu_crash with
+	| None ->
 	(* At this point we expect qemu to outlive us; we will never call waitpid *)
 	Forkhelpers.dontwaitpid qemu_pid
+	| Some _ ->
+	(* We register a callback to be run asynchronously in case qemu fails/crashes or is killed *)
+	let waitpid_async x ~callback =
+		ignore(Thread.create (fun x->callback (try Forkhelpers.waitpid_fail_if_bad_exit x;None with e->Some e)) x)
+	in
+	waitpid_async qemu_pid ~callback:(fun qemu_crash -> Forkhelpers.(
+		debug "Finished waiting qemu pid=%d for domid=%d" (getpid qemu_pid) domid;
+		let crash_reason = match qemu_crash with
+		| None -> debug "domid=%d qemu-pid=%d: detected exit=0" domid (getpid qemu_pid);
+			"exit:0"
+		| Some e -> (match e with
```

Comment:
I'd probably use `begin..end` here around the new `match` instead of nesting additional parenthesis